File: src/app/sim/arena/arena.ts
import { Edge } from './edge'
import { Node } from './node'

export class Arena {

  nodes: Node[] = []
  edges: Edge[] = []

  static randomGrid(width: number, height: number): Arena {
    const arena = new Arena()
    arena.nodes = new Array(width * height)

    // Construct each arena node
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const i = y * width + x
        arena.nodes[i] = new Node(i, x, y)
      }
    }

    // Link them together in a grid
    for (let y = 0; y < height; y++) {
      for (let x = 0; x < width; x++) {
        const i = y * width + x
        const downi = (y + 1) * width + x
        const righti = y * width + (x + 1)

        const node = arena.nodes[i]

        if (y + 1 < height) {
          const downNode = arena.nodes[downi]
          arena.edges.push(node.link(downNode))
        }

        if (x + 1 < width) {
          const rightNode = arena.nodes[righti]
          arena.edges.push(node.link(rightNode))
        }
      }
    }

    return arena
  }


}

File: arena/arena.go
package arena

import (
	"fmt"
	"strings"
)

// Arena is a graph, a set of nodes and edges.
type Arena struct {
	Width  uint
	Height uint

	Nodes []*Node
	Edges []*Edge
}

func link(a, b *Node) *Edge {
	edge := &Edge{A: a, B: b}

	a.edges = append(a.edges, edge)
	b.edges = append(b.edges, edge)

	return edge
}

// ToDot returns a Graphviz DOT file representing the graph.
func (a *Arena) ToDot() string {
	var builder strings.Builder

	builder.WriteString("graph {\n")

	for _, node := range a.Nodes {
		builder.WriteString(fmt.Sprintf("  N%d [pos=\"%d,-%d!\" shape=\"circle\"];\n",
			node.ID, node.X, node.Y))
	}

	builder.WriteString("\n")

	for _, edge := range a.Edges {
		builder.WriteString(fmt.Sprintf("  N%d -- N%d [weight=%d label=\"%d\"];\n",
			edge.A.ID, edge.B.ID, edge.Weight, edge.Weight))
	}

	builder.WriteString("}\n")

	return builder.String()
}
